orderbook/src/solvable_orders.rs

```diff
@@ -0,0 +1,379 @@
+            let balance = match balance {
+                Ok(balance) => balance,
+                Err(err) => {
+                    tracing::warn!(
+                        owner = %query.owner,
+                        token = %query.token,
+                        source = ?query.source,
+                        error = ?err,
+                        "failed to get balance"
+                    );
+                    continue;
+                }
+            };
+            new_balances.insert(query, balance);
+        }
+
+        let mut orders = solvable_orders(orders, &new_balances);
+        for order in &mut orders {
+            let query = Query::from_order(order);
+            order.order_meta_data.available_balance = new_balances.get(&query).copied();
+        }
+
+        *self.cache.lock().unwrap() = Inner {
+            orders,
+            balances: new_balances,
+            block,
+            update_time: Instant::now(),
+        };
+
+        Ok(())
+    }
+}
+
+/// Returns existing balances and Vec of queries that need to be peformed.
+fn new_balances(old_balances: &Balances, orders: &[Order]) -> (HashMap<Query, U256>, Vec<Query>) {
+    let mut new_balances = HashMap::new();
+    let mut missing_queries = HashSet::new();
+    for order in orders {
+        let query = Query::from_order(order);
+        match old_balances.get(&query) {
+            Some(balance) => {
+                new_balances.insert(query, *balance);
+            }
+            None => {
+                missing_queries.insert(query);
+            }
+        }
+    }
+    let missing_queries = Vec::from_iter(missing_queries);
+    (new_balances, missing_queries)
+}
+
+// The order book has to make a choice for which orders to include when a user has multiple orders
+// selling the same token but not enough balance for all of them.
+// Assumes balance fetcher is already tracking all balances.
+fn solvable_orders(mut orders: Vec<Order>, balances: &Balances) -> Vec<Order> {
+    let mut orders_map = HashMap::<Query, Vec<Order>>::new();
+    orders.sort_by_key(|order| std::cmp::Reverse(order.order_meta_data.creation_date));
+    for order in orders {
+        let key = Query::from_order(&order);
+        orders_map.entry(key).or_default().push(order);
+    }
+
+    let mut result = Vec::new();
+    for (key, orders) in orders_map {
+        let mut remaining_balance = match balances.get(&key) {
+            Some(balance) => *balance,
+            None => continue,
+        };
+        for order in orders {
+            // TODO: This is overly pessimistic for partially filled orders where the needed balance
+            // is lower. For partially fillable orders that cannot be fully filled because of the
+            // balance we could also give them as much balance as possible instead of skipping. For
+            // that we first need a way to communicate this to the solver. We could repurpose
+            // availableBalance for this.
```

Comment:
Since this comment is related to the "epic" task of supporting partially fillable orders perhaps we should directly link it to the issue we have for it.

https://github.com/gnosis/gp-v2-services/issues/673

(or vice versa).

Reply:
When I merge this I'll add a comment to the issue.